# Data Profiling — Online Retail Dataset

## Objective

The objective of this notebook is to understand the structure, quality, and characteristics of the raw Online Retail dataset before performing any data cleaning or transformation.

## Profiling Scope

This notebook covers:

* Dataset structure and dimensions
* Column names and data types
* Missing values
* Duplicate rows
* Unique values
* Descriptive statistics
* Potential invalid values
* Potential outliers
* Initial data quality observations
* Business interpretation

> **Note:** No data cleaning or transformation is performed in this notebook.

## 1. Import Libraries

In [3]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [4]:
pd.set_option("display.max_columns", None)
pd.set_option("display.max_rows", 100)

sns.set_theme(style="whitegrid")

## 2. Load Dataset

In [7]:
file_path = "../data/raw/online_retail.xlsx"

df = pd.read_excel(file_path)

In [8]:
df.head()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
0,536365,85123A,WHITE HANGING HEART T-LIGHT HOLDER,6,2010-12-01 08:26:00,2.55,17850.0,United Kingdom
1,536365,71053,WHITE METAL LANTERN,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
2,536365,84406B,CREAM CUPID HEARTS COAT HANGER,8,2010-12-01 08:26:00,2.75,17850.0,United Kingdom
3,536365,84029G,KNITTED UNION FLAG HOT WATER BOTTLE,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom
4,536365,84029E,RED WOOLLY HOTTIE WHITE HEART.,6,2010-12-01 08:26:00,3.39,17850.0,United Kingdom


In [9]:
df.tail()

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
541904,581587,22613,PACK OF 20 SPACEBOY NAPKINS,12,2011-12-09 12:50:00,0.85,12680.0,France
541905,581587,22899,CHILDREN'S APRON DOLLY GIRL,6,2011-12-09 12:50:00,2.10,12680.0,France
541906,581587,23254,CHILDRENS CUTLERY DOLLY GIRL,4,2011-12-09 12:50:00,4.15,12680.0,France
541907,581587,23255,CHILDRENS CUTLERY CIRCUS PARADE,4,2011-12-09 12:50:00,4.15,12680.0,France
541908,581587,22138,BAKING SET 9 PIECE RETROSPOT,3,2011-12-09 12:50:00,4.95,12680.0,France


## 3. Initial Dataset Inspection

In [17]:
df.shape

(541909, 8)

In [11]:
print(f"Rows    : {df.shape[0]:,}")
print(f"Columns : {df.shape[1]}")

Rows    : 541,909
Columns : 8


In [12]:
df.columns.tolist()

['InvoiceNo',
 'StockCode',
 'Description',
 'Quantity',
 'InvoiceDate',
 'UnitPrice',
 'CustomerID',
 'Country']

In [13]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 541909 entries, 0 to 541908
Data columns (total 8 columns):
 #   Column       Non-Null Count   Dtype         
---  ------       --------------   -----         
 0   InvoiceNo    541909 non-null  object        
 1   StockCode    541909 non-null  object        
 2   Description  540455 non-null  object        
 3   Quantity     541909 non-null  int64         
 4   InvoiceDate  541909 non-null  datetime64[us]
 5   UnitPrice    541909 non-null  float64       
 6   CustomerID   406829 non-null  float64       
 7   Country      541909 non-null  str           
dtypes: datetime64[us](1), float64(2), int64(1), object(3), str(1)
memory usage: 33.1+ MB


In [14]:
df.describe()

,Quantity,InvoiceDate,UnitPrice,CustomerID
count,541909.000000,541909,541909.000000,406829.000000
mean,9.552250,2011-07-04 13:34:57.156386,4.611114,15287.690570
min,-80995.000000,2010-12-01 08:26:00,-11062.060000,12346.000000
25%,1.000000,2011-03-28 11:34:00,1.250000,13953.000000
50%,3.000000,2011-07-19 17:17:00,2.080000,15152.000000
75%,10.000000,2011-10-19 11:27:00,4.130000,16791.000000
max,80995.000000,2011-12-09 12:50:00,38970.000000,18287.000000
std,218.081158,NaN,96.759853,1713.600303


In [20]:
df.describe(include="object")

C:\Users\user\AppData\Local\Temp\ipykernel_28564\702825166.py:1: Pandas4Warning: For backward compatibility, 'str' dtypes are included by select_dtypes when 'object' dtype is specified. This behavior is deprecated and will be removed in a future version. Explicitly pass 'str' to `include` to select them, or to `exclude` to remove them and silence this warning.
See https://pandas.pydata.org/docs/user_guide/migration-3-strings.html#string-migration-select-dtypes for details on how to write code that works with pandas 2 and 3.
  df.describe(include="object")


,InvoiceNo,StockCode,Description,Country
count,541909,541909,540455,541909
unique,25900,4070,4223,38
top,573585,85123A,WHITE HANGING HEART T-LIGHT HOLDER,United Kingdom
freq,1114,2313,2369,495478


## Initial Data Profiling Findings

The Online Retail dataset contains **541,909 transaction-line records across 8 columns**. The dataset covers transaction, product, customer, and geographic information.

### Key Findings

1. **Transaction-level structure**
   The dataset contains transaction-line records rather than one row per order. There are 25,900 unique invoice numbers, indicating that a single invoice can contain multiple product lines.

2. **Customer identification**
   `CustomerID` contains 406,829 non-null values out of 541,909 records. Approximately **24.93% of records do not have a CustomerID**. This is an important data quality issue because customer identification is required for customer-level analysis and RFM segmentation.

3. **Missing product descriptions**
   `Description` contains 1,454 missing values, representing approximately **0.27% of all records**.

4. **Transaction dates**
   `InvoiceDate` is already stored as a datetime data type. The observed transaction period ranges from December 2010 to December 2011.

5. **Quantity anomalies**
   `Quantity` contains negative values and extreme observations. The minimum value is -80,995 and the maximum is 80,995. These records require further investigation because negative quantities may represent cancellations or returns rather than data-entry errors.

6. **Unit price anomalies**
   `UnitPrice` contains negative and highly extreme values. The median unit price is 2.08, while the maximum reaches 38,970. These observations require further investigation before determining an appropriate treatment.

7. **Geographic concentration**
   The dataset contains 38 countries, but the majority of transaction records are associated with the United Kingdom. This indicates a strong concentration of business activity in the UK market.

### Initial Business Implications

The dataset is suitable for customer-level analysis and RFM segmentation, but several data quality considerations must be addressed before analysis. In particular, missing `CustomerID` values, negative quantities, unusual unit prices, and transaction cancellations need to be investigated before creating the final analytical dataset.

## 4. Missing Value Analysis

In [21]:
missing_summary = pd.DataFrame({
    "missing_count": df.isna().sum(),
    "missing_percentage": df.isna().mean() * 100
})

missing_summary = missing_summary.sort_values(
    "missing_count",
    ascending=False
)

missing_summary["missing_percentage"] = (
    missing_summary["missing_percentage"].round(2)
)

missing_summary

,missing_count,missing_percentage
CustomerID,135080,24.93
Description,1454,0.27
StockCode,0,0.00
InvoiceNo,0,0.00
Quantity,0,0.00
InvoiceDate,0,0.00
UnitPrice,0,0.00
Country,0,0.00


### Missing Value Findings

The dataset contains missing values in two columns: `CustomerID` and `Description`.

* `CustomerID` has **135,080 missing values (24.93%)**, making it the most significant missing-data issue in the dataset. Since customer identification is required for customer-level analysis and RFM segmentation, records without a `CustomerID` cannot be reliably assigned to a customer.
* `Description` has **1,454 missing values (0.27%)**. The missing proportion is relatively small, and `StockCode` is complete across all records, so these records require further investigation before deciding how they should be treated.
* The remaining six columns contain no missing values.

### Business Implication

Missing `CustomerID` values may reduce the number of transactions available for customer-level analysis. However, these records should not be removed during the profiling stage. Their relationship with transaction cancellations, product information, and other transaction characteristics should first be investigated during the data cleaning phase.

In [22]:
customer_missing = df["CustomerID"].isna()

cancelled = df["InvoiceNo"].astype(str).str.startswith("C")

comparison = pd.DataFrame({
    "CustomerID_Status": [
        "Missing",
        "Available"
    ],
    "Rows": [
        customer_missing.sum(),
        (~customer_missing).sum()
    ],
    "Cancelled_Rows": [
        (customer_missing & cancelled).sum(),
        ((~customer_missing) & cancelled).sum()
    ]
})

comparison

,CustomerID_Status,Rows,Cancelled_Rows
0,Missing,135080,383
1,Available,406829,8905


In [23]:
comparison["Cancellation_Rate"] = (
    comparison["Cancelled_Rows"] /
    comparison["Rows"] * 100
).round(2)

comparison

,CustomerID_Status,Rows,Cancelled_Rows,Cancellation_Rate
0,Missing,135080,383,0.28
1,Available,406829,8905,2.19


In [24]:
df.groupby(
    df["CustomerID"].isna()
)["Quantity"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
CustomerID,,,,,
False,406829,12.061303,5.0,-80995,80995
True,135080,1.995573,1.0,-9600,5568


In [25]:
df.groupby(
    df["CustomerID"].isna()
)["Country"].nunique()

CustomerID
False    37
True      9
Name: Country, dtype: int64

### CustomerID Missingness Investigation

Further investigation shows that transactions with and without `CustomerID` have different characteristics.

Transactions with an available `CustomerID` have a cancellation rate of **2.19%**, compared with **0.28%** for transactions without a `CustomerID`. This indicates that missing customer identification is not evenly distributed across transaction types.

The quantity distribution also differs substantially. Transactions with an available `CustomerID` have a median quantity of **5 units**, while transactions without a `CustomerID` have a median quantity of **1 unit**. Extreme quantity values are also more pronounced among records with an identified customer.

Geographically, transactions with an available `CustomerID` span **37 countries**, while transactions without a `CustomerID` span only **9 countries**.

These findings suggest that records with missing `CustomerID` may have different transaction characteristics. Therefore, missing `CustomerID` records should be investigated further rather than treated as random missing data.

For customer-level analysis and RFM segmentation, only transactions that can be reliably associated with a customer will be eligible for customer-level aggregation. However, the raw records will remain unchanged until the data cleaning phase.

In [26]:
missing_customer_country = (
    df.loc[df["CustomerID"].isna(), "Country"]
    .value_counts()
)

missing_customer_country

Country
United Kingdom    133600
EIRE                 711
Hong Kong            288
Unspecified          202
Switzerland          125
France                66
Israel                47
Portugal              39
Bahrain                2
Name: count, dtype: int64

In [27]:
country_customer_summary = (
    df.groupby("Country")
      .agg(
          total_rows=("CustomerID", "size"),
          missing_customer_id=("CustomerID", lambda x: x.isna().sum())
      )
)

country_customer_summary["missing_percentage"] = (
    country_customer_summary["missing_customer_id"]
    / country_customer_summary["total_rows"]
    * 100
).round(2)

country_customer_summary.sort_values(
    "missing_percentage",
    ascending=False
)

,total_rows,missing_customer_id,missing_percentage
Country,,,
Hong Kong,288,288,100.00
Unspecified,446,202,45.29
United Kingdom,495478,133600,26.96
Israel,297,47,15.82
Bahrain,19,2,10.53
EIRE,8196,711,8.67
Switzerland,2002,125,6.24
Portugal,1519,39,2.57
France,8557,66,0.77


#### Missing CustomerID by Country

Missing `CustomerID` values are highly concentrated geographically.

The United Kingdom contributes **133,600 missing CustomerID records**, representing the vast majority of all missing customer identifiers. However, the missing rate within the UK is **26.96%**, indicating that more than one-quarter of UK transaction records cannot be directly linked to a customer.

Hong Kong shows a different pattern: all **288 transaction records** from Hong Kong have a missing `CustomerID`, resulting in a **100% missing rate**.

The `Unspecified` country category also has a relatively high missing rate of **45.29%**.

These patterns indicate that missing `CustomerID` values are not randomly distributed across countries. This is important when interpreting customer-level analysis because the resulting customer population may not fully represent all transaction activity.

For RFM and customer segmentation, transactions without a valid `CustomerID` will not be eligible for customer-level aggregation. However, these records will remain available for transaction-level analysis where appropriate.

## 5. Duplicate Row Analysis

In [28]:
duplicate_count = df.duplicated().sum()

print(f"Duplicate rows: {duplicate_count:,}")

Duplicate rows: 5,268


In [29]:
duplicate_percentage = (
    duplicate_count / len(df) * 100
)

print(f"Duplicate percentage: {duplicate_percentage:.2f}%")

Duplicate percentage: 0.97%


In [30]:
df[df.duplicated()].head(10)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
517,536409,21866,UNION JACK FLAG LUGGAGE TAG,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
527,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
537,536409,22900,SET 2 TEA TOWELS I LOVE LONDON,1,2010-12-01 11:45:00,2.95,17908.0,United Kingdom
539,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
555,536412,22327,ROUND SNACK BOXES SET OF 4 SKULLS,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
587,536412,22273,FELTCRAFT DOLL MOLLY,1,2010-12-01 11:49:00,2.95,17920.0,United Kingdom
589,536412,22749,FELTCRAFT PRINCESS CHARLOTTE DOLL,1,2010-12-01 11:49:00,3.75,17920.0,United Kingdom
594,536412,22141,CHRISTMAS CRAFT TREE TOP ANGEL,1,2010-12-01 11:49:00,2.10,17920.0,United Kingdom
598,536412,21448,12 DAISY PEGS IN WOOD BOX,1,2010-12-01 11:49:00,1.65,17920.0,United Kingdom
600,536412,22569,FELTCRAFT CUSHION BUTTERFLY,2,2010-12-01 11:49:00,3.75,17920.0,United Kingdom


## Duplicate Row Findings

The dataset contains **5,268 exact duplicate rows**, representing approximately **0.97% of all records**.

The duplicate detection was performed across all eight columns, meaning that these records have identical invoice, product, quantity, date, price, customer, and country information.

However, an exact duplicate at the row level does not automatically indicate an invalid transaction. Multiple identical transaction lines may potentially represent repeated purchases or duplicated records in the source data.

Therefore, duplicate records will not be removed during the profiling stage. Further investigation is required to determine whether the duplicates represent legitimate transaction lines or data duplication before defining the cleaning rule.

In [31]:
duplicate_rows = df[df.duplicated(keep=False)]

duplicate_invoice_summary = (
    duplicate_rows
    .groupby("InvoiceNo")
    .size()
    .sort_values(ascending=False)
)

duplicate_invoice_summary.head(20)

InvoiceNo
547651    57
579458    53
577482    52
572103    50
574481    49
552285    48
537781    47
555524    40
575117    40
569897    40
580305    39
578781    39
578921    38
578041    36
578998    34
536796    31
543040    30
537224    30
547498    29
577057    29
dtype: int64

In [32]:
duplicate_invoice_count = duplicate_rows["InvoiceNo"].nunique()

print(
    f"Invoices containing duplicate rows: "
    f"{duplicate_invoice_count:,}"
)

Invoices containing duplicate rows: 1,933


In [33]:
example_invoice = duplicate_rows["InvoiceNo"].iloc[0]

df[df["InvoiceNo"] == example_invoice]

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
483,536409,90199C,5 STRAND GLASS NECKLACE CRYSTAL,3,2010-12-01 11:45:00,6.35,17908.0,United Kingdom
484,536409,21479,WHITE SKULL HOT WATER BOTTLE,1,2010-12-01 11:45:00,3.75,17908.0,United Kingdom
485,536409,22111,SCOTTIE DOG HOT WATER BOTTLE,1,2010-12-01 11:45:00,4.95,17908.0,United Kingdom
486,536409,22785,SQUARECUSHION COVER PINK UNION FLAG,1,2010-12-01 11:45:00,6.75,17908.0,United Kingdom
487,536409,22975,SPACEBOY CHILDRENS EGG CUP,1,2010-12-01 11:45:00,1.25,17908.0,United Kingdom
488,536409,22972,CHILDREN'S SPACEBOY MUG,1,2010-12-01 11:45:00,1.65,17908.0,United Kingdom
489,536409,22866,HAND WARMER SCOTTY DOG DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
490,536409,22568,FELTCRAFT CUSHION OWL,1,2010-12-01 11:45:00,3.75,17908.0,United Kingdom
491,536409,85116,BLACK CANDELABRA T-LIGHT HOLDER,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom
492,536409,22664,TOY TIDY DOLLY GIRL DESIGN,1,2010-12-01 11:45:00,2.10,17908.0,United Kingdom


### Duplicate Row Findings

The dataset contains **5,268 exact duplicate rows**, representing **0.97% of all records**. These duplicate rows are distributed across **1,933 invoices**.

Duplicate detection was performed across all eight columns, meaning that a record is considered an exact duplicate only when all transaction attributes are identical.

Further inspection shows that duplicate records can occur within the same invoice alongside other legitimate product lines. Therefore, duplicate rows should not be interpreted as duplicate invoices.

The presence of exact duplicate transaction lines may cause transaction quantities and revenue to be overstated if retained during aggregation. Therefore, exact duplicates will be removed during the data cleaning phase. However, legitimate repeated purchases with different transaction attributes will be preserved.

In [34]:
duplicate_rows = df[df.duplicated(keep=False)].copy()

duplicate_rows["is_cancelled"] = (
    duplicate_rows["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

duplicate_rows["is_cancelled"].value_counts()

is_cancelled
False    10078
True        69
Name: count, dtype: int64

In [35]:
duplicate_rows.groupby("is_cancelled")["Quantity"].agg(
    ["count", "sum", "mean", "median", "min", "max"]
)

,count,sum,mean,median,min,max
is_cancelled,,,,,,
False,10078,30918,3.067871,1.0,1,1440
True,69,-4012,-58.144928,-1.0,-1296,-1


### Duplicate Investigation

Further investigation shows that duplicate transaction lines are predominantly associated with regular transactions rather than cancellations.

Among the **10,147 records belonging to duplicate groups**, 10,078 records (99.32%) are associated with regular invoices, while only 69 records (0.68%) are associated with cancellation invoices.

The quantity distribution also follows the expected transaction pattern. Regular duplicate-group records have a median quantity of 1 and a minimum quantity of 1, while cancellation duplicate-group records have negative quantities, with a median of -1.

These findings support treating exact duplicate rows as potential duplicate records rather than legitimate repeated transactions when all transaction attributes are identical. Exact duplicates will therefore be removed during the data cleaning phase, while transaction lines with different attributes will be preserved.

In [36]:
unique_summary = pd.DataFrame({
    "unique_count": df.nunique(dropna=True),
    "uniqueness_percentage": (
        df.nunique(dropna=True) / len(df) * 100
    )
})

unique_summary["uniqueness_percentage"] = (
    unique_summary["uniqueness_percentage"].round(2)
)

unique_summary.sort_values(
    "unique_count",
    ascending=False
)

,unique_count,uniqueness_percentage
InvoiceNo,25900,4.78
InvoiceDate,23260,4.29
CustomerID,4372,0.81
Description,4223,0.78
StockCode,4070,0.75
UnitPrice,1630,0.30
Quantity,722,0.13
Country,38,0.01


In [37]:
for col in df.columns:
    print(f"\n{'=' * 50}")
    print(f"{col}")
    print(f"{'=' * 50}")
    print(df[col].nunique(dropna=True), "unique values")
    print(df[col].dropna().unique()[:20])


InvoiceNo
25900 unique values
[536365 536366 536367 536368 536369 536370 536371 536372 536373 536374
 536375 536376 536377 536378 536380 536381 'C536379' 536382 'C536383'
 536384]

StockCode
4070 unique values
['85123A' 71053 '84406B' '84029G' '84029E' 22752 21730 22633 22632 84879
 22745 22748 22749 22310 84969 22623 22622 21754 21755 21777]

Description
4223 unique values
['WHITE HANGING HEART T-LIGHT HOLDER' 'WHITE METAL LANTERN'
 'CREAM CUPID HEARTS COAT HANGER' 'KNITTED UNION FLAG HOT WATER BOTTLE'
 'RED WOOLLY HOTTIE WHITE HEART.' 'SET 7 BABUSHKA NESTING BOXES'
 'GLASS STAR FROSTED T-LIGHT HOLDER' 'HAND WARMER UNION JACK'
 'HAND WARMER RED POLKA DOT' 'ASSORTED COLOUR BIRD ORNAMENT'
 "POPPY'S PLAYHOUSE BEDROOM " "POPPY'S PLAYHOUSE KITCHEN"
 'FELTCRAFT PRINCESS CHARLOTTE DOLL' 'IVORY KNITTED MUG COSY '
 'BOX OF 6 ASSORTED COLOUR TEASPOONS' 'BOX OF VINTAGE JIGSAW BLOCKS '
 'BOX OF VINTAGE ALPHABET BLOCKS' 'HOME BUILDING BLOCK WORD'
 'LOVE BUILDING BLOCK WORD' 'RECIPE BOX WITH METAL

## Unique Value Analysis

The dataset contains different levels of cardinality across its columns, reflecting their respective business roles.

`InvoiceNo` contains **25,900 unique invoice identifiers**, while `CustomerID` contains **4,372 unique customers** among non-null records. This confirms that the dataset is structured at the transaction-line level, with multiple product lines potentially belonging to the same invoice and customer.

`StockCode` contains **4,070 unique product identifiers**, while `Description` contains **4,223 unique descriptions**. The difference in cardinality suggests that some stock codes may be associated with multiple product descriptions and requires further validation.

`InvoiceDate` contains **23,260 unique timestamps**, which is expected because multiple transaction lines can share the same invoice timestamp.

`UnitPrice` and `Quantity` contain 1,630 and 722 unique values respectively. Their cardinality is reasonable for numeric transaction attributes, although their distributions and extreme values require separate outlier analysis.

`Country` contains **38 unique countries**, making it suitable for geographic grouping and market-level analysis.

Overall, the cardinality of the dataset is consistent with a transaction-level retail dataset. `InvoiceNo`, `StockCode`, and `CustomerID` will serve as important identifiers for subsequent transaction, product, and customer-level analysis.

In [38]:
stockcode_description = (
    df.groupby("StockCode")["Description"]
      .nunique(dropna=True)
      .sort_values(ascending=False)
)

stockcode_description.head(20)

StockCode
20713     8
23084     7
21830     6
85175     6
72807A    5
85172     5
23343     5
21181     5
23131     5
85185B    4
85123A    4
46000S    4
72802A    4
72807B    4
22837     4
23196     4
22812     4
22501     4
22121     4
21823     4
Name: Description, dtype: int64

In [39]:
multiple_descriptions = (
    stockcode_description > 1
).sum()

print(
    f"StockCodes with multiple descriptions: "
    f"{multiple_descriptions:,}"
)

StockCodes with multiple descriptions: 650


In [40]:
stockcode_description[
    stockcode_description > 1
].head(20)

StockCode
20713     8
23084     7
21830     6
85175     6
72807A    5
85172     5
23343     5
21181     5
23131     5
85185B    4
85123A    4
46000S    4
72802A    4
72807B    4
22837     4
23196     4
22812     4
22501     4
22121     4
21823     4
Name: Description, dtype: int64

### StockCode and Description Consistency

Further investigation shows that **650 out of 4,070 StockCodes (15.97%) are associated with more than one product description**. Some StockCodes have several different descriptions, with the highest example having eight descriptions.

This indicates that `Description` should not be treated as the primary product identifier because the same `StockCode` can appear with different descriptions.

For subsequent product-level analysis, `StockCode` will therefore be treated as the primary product identifier, while `Description` will be used as a descriptive attribute rather than a unique key.

The variation in descriptions may be caused by changes in product naming, formatting inconsistencies, or other source-data characteristics. These variations do not prevent product-level analysis as long as `StockCode` is used as the product identifier.

## 6. Outlier Analysis

In [41]:
Q1_quantity = df["Quantity"].quantile(0.25)
Q3_quantity = df["Quantity"].quantile(0.75)

IQR_quantity = Q3_quantity - Q1_quantity

lower_quantity = Q1_quantity - 1.5 * IQR_quantity
upper_quantity = Q3_quantity + 1.5 * IQR_quantity

print(f"Q1       : {Q1_quantity}")
print(f"Q3       : {Q3_quantity}")
print(f"IQR      : {IQR_quantity}")
print(f"Lower    : {lower_quantity}")
print(f"Upper    : {upper_quantity}")

Q1       : 1.0
Q3       : 10.0
IQR      : 9.0
Lower    : -12.5
Upper    : 23.5


In [42]:
quantity_outliers = df[
    (df["Quantity"] < lower_quantity) |
    (df["Quantity"] > upper_quantity)
]

print(f"Outlier rows: {len(quantity_outliers):,}")
print(
    f"Outlier percentage: "
    f"{len(quantity_outliers) / len(df) * 100:.2f}%"
)

Outlier rows: 58,619
Outlier percentage: 10.82%


In [43]:
quantity_outliers["Quantity"].describe()

count    58619.000000
mean        50.590969
std        661.524902
min     -80995.000000
25%         24.000000
50%         25.000000
75%         48.000000
max      80995.000000
Name: Quantity, dtype: float64

In [44]:
quantity_outliers = df[
    (df["Quantity"] < lower_quantity) |
    (df["Quantity"] > upper_quantity)
].copy()

quantity_outliers["is_cancelled"] = (
    quantity_outliers["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

quantity_outliers["outlier_type"] = np.where(
    quantity_outliers["Quantity"] < lower_quantity,
    "Negative",
    "Positive"
)

In [45]:
quantity_outliers.groupby(
    ["outlier_type", "is_cancelled"]
)["Quantity"].agg(
    ["count", "sum", "mean", "median", "min", "max"]
)

count      sum        mean  median    min    max
outlier_type is_cancelled                                                  
Negative     False           903  -204757 -226.751938   -56.0  -9600    -13
             True           1081  -252345 -233.436633   -30.0 -80995    -13
Positive     False         56635  3422694   60.434254    25.0     24  80995

In [46]:
quantity_outliers[
    quantity_outliers["Quantity"] > 100
]["Quantity"].describe()

count     4950.000000
mean       298.924242
std       1593.635096
min        101.000000
25%        144.000000
50%        192.000000
75%        288.000000
max      80995.000000
Name: Quantity, dtype: float64

In [47]:
quantity_outliers[
    quantity_outliers["Quantity"] > 1000
][[
    "InvoiceNo",
    "StockCode",
    "Description",
    "Quantity",
    "InvoiceDate",
    "UnitPrice",
    "CustomerID",
    "Country"
]].sort_values(
    "Quantity",
    ascending=False
).head(20)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
540421,581483,23843,"PAPER CRAFT , LITTLE BIRDIE",80995,2011-12-09 09:15:00,2.08,16446.0,United Kingdom
61619,541431,23166,MEDIUM CERAMIC TOP STORAGE JAR,74215,2011-01-18 10:01:00,1.04,12346.0,United Kingdom
502122,578841,84826,ASSTD DESIGN 3D PAPER STICKERS,12540,2011-11-25 15:57:00,0.00,13256.0,United Kingdom
74614,542504,37413,NaN,5568,2011-01-28 12:03:00,0.00,NaN,United Kingdom
421632,573008,84077,WORLD WAR 2 GLIDERS ASSTD DESIGNS,4800,2011-10-27 12:26:00,0.21,12901.0,United Kingdom
206121,554868,22197,SMALL POPCORN HOLDER,4300,2011-05-27 10:52:00,0.72,13135.0,United Kingdom
220843,556231,85123A,?,4000,2011-06-09 15:04:00,0.00,NaN,United Kingdom
97432,544612,22053,EMPIRE DESIGN ROSETTE,3906,2011-02-22 10:43:00,0.82,18087.0,United Kingdom
270885,560599,18007,ESSENTIAL BALM 3.5g TIN IN ENVELOPE,3186,2011-07-19 17:04:00,0.06,14609.0,United Kingdom
52711,540815,21108,FAIRY CAKE FLANNEL ASSORTED COLOUR,3114,2011-01-11 12:55:00,2.10,15749.0,United Kingdom


In [48]:
high_quantity_zero_price = df[
    (df["Quantity"] > 1000) &
    (df["UnitPrice"] == 0)
].copy()

print(f"Rows: {len(high_quantity_zero_price):,}")

high_quantity_zero_price[
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].sort_values(
    "Quantity",
    ascending=False
).head(20)

Rows: 9


,InvoiceNo,StockCode,Description,Quantity,UnitPrice,CustomerID,Country
502122,578841,84826,ASSTD DESIGN 3D PAPER STICKERS,12540,0.0,13256.0,United Kingdom
74614,542504,37413,NaN,5568,0.0,NaN,United Kingdom
220843,556231,85123A,?,4000,0.0,NaN,United Kingdom
263885,560040,23343,came coded as 20713,3100,0.0,NaN,United Kingdom
115807,546139,84988,?,3000,0.0,NaN,United Kingdom
74615,542505,79063D,NaN,2560,0.0,NaN,United Kingdom
160541,550460,47556B,did a credit and did not tick ret,1300,0.0,NaN,United Kingdom
203751,554550,47566B,incorrectly credited C550456 see 47,1300,0.0,NaN,United Kingdom
82795,543258,84611B,NaN,1287,0.0,NaN,United Kingdom


In [49]:
high_quantity_missing_customer = df[
    (df["Quantity"] > 1000) &
    (df["CustomerID"].isna())
]

print(
    f"Quantity > 1000 with missing CustomerID: "
    f"{len(high_quantity_missing_customer):,}"
)

Quantity > 1000 with missing CustomerID: 10


In [50]:
negative_non_cancelled = df[
    (df["Quantity"] < lower_quantity) &
    (~df["InvoiceNo"].astype(str).str.startswith("C"))
]

print(
    f"Negative outliers without cancellation prefix: "
    f"{len(negative_non_cancelled):,}"
)

negative_non_cancelled[
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "InvoiceDate",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].sort_values(
    "Quantity"
).head(20)

Negative outliers without cancellation prefix: 903


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
225529,556690,23005,printing smudges/thrown away,-9600,2011-06-14 10:37:00,0.0,NaN,United Kingdom
225530,556691,23005,printing smudges/thrown away,-9600,2011-06-14 10:37:00,0.0,NaN,United Kingdom
225528,556687,23003,Printing smudges/thrown away,-9058,2011-06-14 10:36:00,0.0,NaN,United Kingdom
115818,546152,72140F,throw away,-5368,2011-03-09 17:25:00,0.0,NaN,United Kingdom
431381,573596,79323W,"Unsaleable, destroyed.",-4830,2011-10-31 15:17:00,0.0,NaN,United Kingdom
341601,566768,16045,NaN,-3667,2011-09-14 17:53:00,0.0,NaN,United Kingdom
323458,565304,16259,NaN,-3167,2011-09-02 12:18:00,0.0,NaN,United Kingdom
263884,560039,20713,wrongly marked. 23343 in box,-3100,2011-07-14 14:27:00,0.0,NaN,United Kingdom
113580,545990,84598,check,-3000,2011-03-08 13:07:00,0.0,NaN,United Kingdom
375429,569466,23270,incorrect stock entry.,-2880,2011-10-04 11:42:00,0.0,NaN,United Kingdom


### Quantity Outlier Findings

Using the IQR method, the lower and upper bounds for `Quantity` were calculated as **-12.5** and **23.5**, respectively. This identified **58,619 rows (10.82%)** as statistical outliers.

However, the statistical definition alone is not sufficient to determine whether these records are invalid. Further investigation revealed several distinct business patterns.

First, 56,635 positive outlier records were identified. Many of these represent high-volume purchases and may be legitimate bulk transactions. Therefore, high positive quantities will not automatically be removed.

Second, 1,984 negative quantity outliers were identified. Of these, 1,081 records are associated with invoices beginning with `C`, indicating cancellation or return-related transactions. Another 903 negative outliers do not have the `C` prefix. Their descriptions include terms such as "thrown away", "unsaleable", "destroyed", "mouldy", and "incorrect stock entry", indicating that these records are more likely to represent inventory adjustments, damaged goods, disposal, or stock corrections rather than customer cancellations.

Third, nine transactions were identified with both `Quantity > 1,000` and `UnitPrice = 0`. Several of these records have missing CustomerID or descriptions containing operational notes such as "incorrectly credited", "came coded as", or "?". These records are strong candidates for non-sales or data-quality records.

Therefore, IQR-based filtering will **not** be used to automatically remove Quantity outliers. Instead, transaction records will be classified based on their business context during the data cleaning stage. Legitimate high-volume purchases will be preserved, while cancellation and non-sales inventory adjustment records will be handled separately.

In [51]:
Q1_price = df["UnitPrice"].quantile(0.25)
Q3_price = df["UnitPrice"].quantile(0.75)

IQR_price = Q3_price - Q1_price

lower_price = Q1_price - 1.5 * IQR_price
upper_price = Q3_price + 1.5 * IQR_price

print(f"Q1       : {Q1_price}")
print(f"Q3       : {Q3_price}")
print(f"IQR      : {IQR_price}")
print(f"Lower    : {lower_price}")
print(f"Upper    : {upper_price}")

Q1       : 1.25
Q3       : 4.13
IQR      : 2.88
Lower    : -3.0700000000000003
Upper    : 8.45


In [52]:
price_outliers = df[
    (df["UnitPrice"] < lower_price) |
    (df["UnitPrice"] > upper_price)
].copy()

print(f"Outlier rows: {len(price_outliers):,}")
print(
    f"Outlier percentage: "
    f"{len(price_outliers) / len(df) * 100:.2f}%"
)

Outlier rows: 39,627
Outlier percentage: 7.31%


In [53]:
price_outliers["UnitPrice"].describe()

count    39627.000000
mean        31.084752
std        356.699161
min     -11062.060000
25%          9.950000
50%         11.630000
75%         15.790000
max      38970.000000
Name: UnitPrice, dtype: float64

In [54]:
negative_price = df[
    df["UnitPrice"] < 0
].copy()

print(f"Negative UnitPrice rows: {len(negative_price):,}")

negative_price[
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "InvoiceDate",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].sort_values(
    "UnitPrice"
).head(20)

Negative UnitPrice rows: 2


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
299983,A563186,B,Adjust bad debt,1,2011-08-12 14:51:00,-11062.06,NaN,United Kingdom
299984,A563187,B,Adjust bad debt,1,2011-08-12 14:52:00,-11062.06,NaN,United Kingdom


In [55]:
negative_price["is_cancelled"] = (
    negative_price["InvoiceNo"]
    .astype(str)
    .str.startswith("C")
)

negative_price.groupby("is_cancelled")["UnitPrice"].agg(
    ["count", "mean", "median", "min", "max"]
)

,count,mean,median,min,max
is_cancelled,,,,,
False,2,-11062.06,-11062.06,-11062.06,-11062.06


In [56]:
extreme_price = df[
    df["UnitPrice"] > 100
].copy()

print(f"UnitPrice > 100: {len(extreme_price):,}")

extreme_price[
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "InvoiceDate",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].sort_values(
    "UnitPrice",
    ascending=False
).head(30)

UnitPrice > 100: 1,036


,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
222681,C556445,M,Manual,-1,2011-06-10 15:31:00,38970.00,15098.0,United Kingdom
524602,C580605,AMAZONFEE,AMAZON FEE,-1,2011-12-05 11:36:00,17836.46,NaN,United Kingdom
43702,C540117,AMAZONFEE,AMAZON FEE,-1,2011-01-05 09:55:00,16888.02,NaN,United Kingdom
43703,C540118,AMAZONFEE,AMAZON FEE,-1,2011-01-05 09:57:00,16453.71,NaN,United Kingdom
15016,C537630,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:04:00,13541.33,NaN,United Kingdom
15017,537632,AMAZONFEE,AMAZON FEE,1,2010-12-07 15:08:00,13541.33,NaN,United Kingdom
16356,C537651,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:49:00,13541.33,NaN,United Kingdom
16232,C537644,AMAZONFEE,AMAZON FEE,-1,2010-12-07 15:34:00,13474.79,NaN,United Kingdom
524601,C580604,AMAZONFEE,AMAZON FEE,-1,2011-12-05 11:35:00,11586.50,NaN,United Kingdom
299982,A563185,B,Adjust bad debt,1,2011-08-12 14:50:00,11062.06,NaN,United Kingdom


In [57]:
zero_price = df[
    df["UnitPrice"] == 0
].copy()

print(f"UnitPrice = 0 rows: {len(zero_price):,}")

UnitPrice = 0 rows: 2,515


In [58]:
zero_price[
    [
        "InvoiceNo",
        "StockCode",
        "Description",
        "Quantity",
        "InvoiceDate",
        "UnitPrice",
        "CustomerID",
        "Country"
    ]
].head(30)

,InvoiceNo,StockCode,Description,Quantity,InvoiceDate,UnitPrice,CustomerID,Country
622,536414,22139,NaN,56,2010-12-01 11:52:00,0.0,NaN,United Kingdom
1970,536545,21134,NaN,1,2010-12-01 14:32:00,0.0,NaN,United Kingdom
1971,536546,22145,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1972,536547,37509,NaN,1,2010-12-01 14:33:00,0.0,NaN,United Kingdom
1987,536549,85226A,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
1988,536550,85044,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2024,536552,20950,NaN,1,2010-12-01 14:34:00,0.0,NaN,United Kingdom
2025,536553,37461,NaN,3,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2026,536554,84670,NaN,23,2010-12-01 14:35:00,0.0,NaN,United Kingdom
2406,536589,21777,NaN,-10,2010-12-01 16:50:00,0.0,NaN,United Kingdom


In [59]:
zero_price["CustomerID"].isna().value_counts()

CustomerID
True     2475
False      40
Name: count, dtype: int64

### UnitPrice Outlier Findings

The IQR method identified an upper bound of **8.45** and classified **39,627 rows (7.31%)** as statistical outliers.

Further investigation showed that these outliers cannot be automatically considered invalid transactions.

Two records have negative `UnitPrice` values of **-11,062.06**. Both records have the description `Adjust bad debt`, missing CustomerID, and are not cancellations. These records represent accounting adjustments rather than normal product sales.

There are also **1,036 records with UnitPrice greater than 100**. Several of the most extreme records correspond to `AMAZONFEE`, `POSTAGE`, `Manual`, and other operational transactions. Therefore, extreme prices may represent fees or accounting-related transactions rather than product prices.

A total of **2,515 records have UnitPrice = 0**. Of these, 2,475 records (98.4%) have missing CustomerID. Many also have missing descriptions, suggesting that a substantial portion may represent non-sales or incomplete transaction records. However, zero-priced transactions should not automatically be removed because they may also represent legitimate free items, promotional items, or other business adjustments.

Therefore, IQR-based filtering will not be used to remove UnitPrice outliers. Instead, transaction-type indicators and business rules will be used to distinguish normal product sales from cancellations, fees, accounting adjustments, and other non-sales records.

## Final Data Quality Summary

| Issue | Finding | Impact | Action |
|---|---|---|---|
| Missing CustomerID | 135,080 rows (24.93%) | Customer-level analysis tidak dapat dilakukan pada transaksi tersebut | Exclude dari RFM dan customer segmentation |
| Missing Description | 1,454 rows (0.27%) | Informasi produk tidak lengkap | Retain jika StockCode valid |
| Duplicate Rows | 5,268 rows (0.97%) | Berpotensi menyebabkan double counting | Remove exact duplicates |
| Cancellation | 8,974 rows (1.66%) | Dapat menyebabkan revenue dan quantity menjadi misleading | Exclude dari normal sales analysis |
| Negative Quantity | 1,984 outlier rows | Dapat merepresentasikan cancellation, return, atau operational adjustment | Investigate berdasarkan konteks transaksi |
| Extreme Positive Quantity | 56,635 rows | Tidak selalu merupakan data error; dapat berupa bulk purchase atau operational transaction | Investigate, tidak langsung dihapus |
| Negative UnitPrice | 2 rows | Bukan normal product sales | Exclude dari normal sales analysis |
| UnitPrice > 100 | 1,036 rows | Sebagian berasal dari AMAZON FEE, POSTAGE, dan Manual transactions | Investigate non-product transactions |
| UnitPrice = 0 | 2,515 rows | Tidak menghasilkan revenue dan dapat mengganggu transaction/customer metrics | Exclude dari normal sales analysis |
| Multiple Description per StockCode | 650 StockCodes | Menunjukkan inkonsistensi product description | Gunakan StockCode sebagai product identifier |

## Business Interpretation

The profiling process shows that the dataset contains not only normal product sales, but also cancellations, returns, operational adjustments, fees, and other non-standard transactions.

### 1. Customer Analysis Limitation

Approximately **24.93% of transactions do not have a CustomerID**.

This means these transactions cannot be reliably attributed to a specific customer. Therefore, they should not be included in customer-level analysis such as:

- RFM analysis
- Customer segmentation
- Customer purchase frequency
- Customer monetary value
- Customer retention analysis

However, transactions without CustomerID may still contain valid information for overall sales and product analysis.

---

### 2. Transaction Quality

The dataset contains **5,268 exact duplicate rows**, representing approximately **0.97% of all records**.

These duplicates may cause double counting when calculating:

- Revenue
- Quantity sold
- Product performance
- Transaction metrics

Exact duplicate rows should therefore be removed during the data cleaning stage.

---

### 3. Cancellation and Returns

Approximately **1.66% of rows are cancellation transactions**.

Cancellation transactions are generally identified by an `InvoiceNo` beginning with `C` and negative quantities.

These transactions should not be treated as normal sales because including them in sales calculations could distort:

- Revenue
- Quantity sold
- Average Order Value
- Product performance
- Customer spending

---

### 4. Quantity Outliers

The IQR method identified **58,619 quantity outlier rows (10.82%)**.

However, investigation shows that extreme quantities are not necessarily data errors. Some may represent:

- Bulk purchases
- Large orders
- Inventory adjustments
- Damaged or unsaleable products
- Incorrect stock entries

Therefore, quantity outliers should **not be removed solely based on statistical thresholds**.

Business context must be considered before deciding whether a transaction is valid.

---

### 5. UnitPrice Outliers

The IQR method identified **39,627 UnitPrice outlier rows (7.31%)**.

Investigation revealed that many extreme prices are associated with non-product transactions such as:

- `AMAZON FEE`
- `POSTAGE`
- `Manual`
- `Adjust bad debt`

Therefore, statistical outliers in UnitPrice cannot automatically be considered invalid product prices.

---

### 6. Zero and Negative UnitPrice

There are **2,515 transactions with UnitPrice = 0** and **2 transactions with negative UnitPrice**.

The negative UnitPrice transactions are associated with `Adjust bad debt`, while many zero-price transactions have missing customer and product descriptions.

These transactions do not represent normal product sales and should be excluded from the valid sales dataset.

---

### 7. Product Identification

There are **650 StockCodes with multiple descriptions**.

This indicates that `Description` is not fully consistent across the dataset.

Therefore, `StockCode` will be treated as the primary product identifier, while `Description` will be used as supporting product information.

---

### Overall Business Interpretation

The dataset is suitable for sales performance, product analysis, customer behavior, and RFM segmentation. However, the raw dataset should not be used directly for business analysis.

The main challenge is distinguishing **valid product sales from cancellations, returns, operational adjustments, fees, and other non-standard transactions**.

Therefore, the next phase will focus on creating explicit business rules for data cleaning rather than simply removing statistical outliers.